rebasebot/lifecycle_hooks.py

```diff
@@ -0,0 +1,171 @@
+            hash_suffix = str(abs(hash(self.script_location)))[:5]  # Hash the script location to avoid name conflicts
+            basename, ext = os.path.splitext(os.path.basename(file_path))
+            self.script_file_path = f"{temp_hook_dir}{basename}-{hash_suffix}{ext}"
+            try:
+                with open(f"{self.script_file_path}", "w", encoding='latin1') as f:
+                    f.write(gitwd.git.show(f"{ref}:{file_path}"))
+            except git.GitCommandError as e:
+                raise ValueError(f"Failed to retrieve script from git reference {ref}") from e
+            os.chmod(f"{self.script_file_path}", 0o755)
+
+    def __str__(self):
+        return self.script_location
+
+    def __call__(self):
+        try:
+            with subprocess.Popen(
+                [self.script_file_path],
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+                text=True
+            ) as process:
+
+                streams = [process.stdout, process.stderr]
+
+                while streams:
+                    # Wait for output from both stdout and stderr
+                    readable, _, _ = select.select(streams, [], [])
+                    for stream in readable:
+                        line = stream.readline()
+                        if line:
+                            if stream == process.stderr:
+                                print(line, file=sys.stderr, end='')
+                            else:
+                                print(line, end='')
+                        else:
+                            # Remove closed stream from the list
+                            streams.remove(stream)
+
+                # Wait for the process to finish
+                return_code = process.poll()
+                if return_code != 0:
+                    raise subprocess.CalledProcessError(return_code, self.script_file_path)
+
+        except subprocess.CalledProcessError as e:
+            print(f"Script failed with error:\n{e}")
+
+
+def _setup_environment_variables(args):
+    """Sets up environment variables with rebasebot parameters for lifecycle hook scripts."""
+    os.environ["REBASEBOT_SOURCE"] = args.source.branch
+    os.environ["REBASEBOT_DEST"] = args.dest.branch
+    os.environ["REBASEBOT_REBASE"] = args.rebase.branch
+    os.environ["REBASEBOT_WORKING_DIR"] = args.working_dir
+    os.environ["REBASEBOT_GIT_USERNAME"] = args.git_username
+    os.environ["REBASEBOT_GIT_EMAIL"] = args.git_email
+
+
+class LifecycleHooks:
+    """LifecycleHooks stores lifecycle hook script a"""
+    hooks: dict[LifecycleHook, list[LifecycleHookScript]] = {}
+    tmp_hook_scripts_dir: str = None
+
+    def attach_script_to_hook(self, hook: LifecycleHook, script: LifecycleHookScript):
+        """Adds a script to the specified hook."""
+        if hook not in self.hooks:
+            self.hooks[hook] = []
+        self.hooks[hook].append(script)
```

Comment:
it seems like we allow multiple scripts per hook, but i didn't see this mentioned in the docs. did i miss it?

Reply:
Mentioned this in docs now.